### PR TITLE
Fix import for master branch

### PR DESCRIPTION
**Pull request by pierallard**
- Generates CSV instead of YML
- Add a CSV writer service
- Correct typo and bugs
- Generate dummy locale file

### CODE REVIEW

File: Generator/AssetCategoryAccessGenerator.php [hunk 1]
```diff
@@ -15,55 +16,45 @@
...
-    protected function writeYamlFile(array $data, $outputDir)
-    {
-        $dumper = new Yaml\Dumper();
-        $yamlData = $dumper->dump($data, 3, 0, true, true);
-
-        file_put_contents($outputDir.'/'.self::ASSET_CATEGORY_ACCESSES_FILENAME, $yamlData);
+        $this->writer
+            ->setFilename($globalConfig['output_dir'] . '/' . self::ASSET_CATEGORY_ACCESSES_FILENAME)
+            ->setData($data)
+            ->write();
```
Comment (nidup) [inline Generator/AssetCategoryAccessGenerator.php]: :+1:

File: Resources/config/internal_jobs.yml [hunk 2]
```diff
@@ -34,3 +34,241 @@ jobs:
             enclosure:  '"'
             withHeader: true
             filePath:   /tmp/products_export_%locale%_%scope%_%datetime%.csv
+    rule_import:
```
Comment (nidup) [inline Resources/config/internal_jobs.yml]: means we always add EE configuration ? in CE too ? does it work?

File: Writer/CsvWriter.php [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+namespace Pim\Bundle\DataGeneratorBundle\Writer;
+
+/**
+ * Write CSV files
+ *
+ * @author    Pierre Allard <pierre.allard@akeneo.com>
+ * @copyright 2016 Akeneo SAS (http://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+class CsvWriter implements WriterInterface
```
Comment (nidup) [inline Writer/CsvWriter.php]: ok, imho, we should use as much as we can existing PIM behavior and finish by using the real complete exports.

To prepare this your CsvWriter should implement https://github.com/akeneo/pim-community-dev/blob/master/src/Akeneo/Component/Batch/Item/ItemWriterInterface.php especially implements "public function write(array $items);", no need of setData() then write()

You just have to add the setFilename() method in your own writer

File: Generator/AssociationTypeGenerator.php [hunk 4]
```diff
@@ -37,21 +50,22 @@ public function generate(array $globalConfig, array $config, ProgressHelper $pro
...
+            $data[] = $this->normalizeAssociationType($associationType);
 
             $progress->advance();
         }
 
-        $data = [self::ASSOCIATIONS => $data];
-
-        $this->writeYamlFile($data, $globalConfig['output_dir']);
+        $this->writer
+            ->setFilename($globalConfig['output_dir'] . '/' . self::ASSOCIATION_TYPES_FILENAME)
```
Comment (arnolanglade) [inline Generator/AssociationTypeGenerator.php]: sprintf ?

File: Generator/AssetCategoryGenerator.php [hunk 5]
```diff
@@ -46,33 +55,13 @@ public function generate(array $globalConfig, array $config, ProgressHelper $pro
             $assetCategories[0][$key] = implode(' ', $faker->words(3));
         }
 
-        $headers = array_keys($assetCategories[0]);
-
-        $this->writeCsvFile($assetCategories, $headers, $globalConfig['output_dir']);
+        $this->writer
+            ->setFilename($globalConfig['output_dir'] . '/' . self::ASSET_CATEGORIES_FILENAME)
```
Comment (arnolanglade) [inline Generator/AssetCategoryGenerator.php]: sprintf ?

File: Generator/AttributeGenerator.php [hunk 6]
```diff
@@ -66,21 +69,17 @@ public function generate(array $globalConfig, array $config, ProgressHelper $pro
...
-        $count = (int) $config['count'];
+        $count = (int)$config['count'];
 
-        $localizableProbability = (float) $config['localizable_probability'];
-        $scopableProbability    = (float) $config['scopable_probability'];
-        $locScopableProbability = (float) $config['localizable_and_scopable_probability'];
-        $gridFilterProbability  = (float) $config['useable_as_grid_filter_probability'];
-        $minVariantAxes         = (int) $config['min_variant_axes'];
-        $minVariantAttributes   = (int) $config['min_variant_attributes'];
+        $localizableProbability = (float)$config['localizable_probability'];
```
Comment (aRn0D) [inline Generator/AttributeGenerator.php]: (float) $config

File: Generator/AttributeGenerator.php [hunk 7]
```diff
@@ -66,21 +69,17 @@ public function generate(array $globalConfig, array $config, ProgressHelper $pro
...
+        $scopableProbability    = (float)$config['scopable_probability'];
+        $locScopableProbability = (float)$config['localizable_and_scopable_probability'];
+        $gridFilterProbability  = (float)$config['useable_as_grid_filter_probability'];
+        $minVariantAxes         = (int)$config['min_variant_axes'];
```
Comment (aRn0D) [inline Generator/AttributeGenerator.php]: (int) $config

File: Generator/AttributeGenerator.php [hunk 8]
```diff
@@ -127,8 +126,8 @@ public function generate(array $globalConfig, array $config, ProgressHelper $pro
                 $attribute['localizable'] = 1;
                 $attribute['scopable'] = 1;
             } else {
-                $attribute['localizable'] = (int) $this->faker->boolean($localizableProbability);
-                $attribute['scopable'] = (int) $this->faker->boolean($scopableProbability);
+                $attribute['localizable'] = (int)$this->faker->boolean($localizableProbability);
```
Comment (aRn0D) [inline Generator/AttributeGenerator.php]: idem

File: Generator/AttributeGenerator.php [hunk 9]
```diff
@@ -144,21 +143,20 @@ public function generate(array $globalConfig, array $config, ProgressHelper $pro
         }
 
         foreach ($this->attributes as $code => $attribute) {
-            foreach ($this->getLocalizedRandomLabels($type) as $localeCode => $label) {
+            foreach ($this->getLocalizedRandomLabels($attribute['type']) as $localeCode => $label) {
```
Comment (arnolanglade) [inline Generator/AttributeGenerator.php:146]: `$localeCode => $localeLabel` or `$code => $label` ?

File: Generator/AttributeGenerator.php [hunk 10]
```diff
@@ -144,21 +143,20 @@ public function generate(array $globalConfig, array $config, ProgressHelper $pro
...
                 $this->attributes[$code]['label-' . $localeCode] = $label;
             }
 
             if (!isset($attribute['useable_as_grid_filter'])) {
-                $useable = (int) $this->faker->boolean($gridFilterProbability);
+                $useable = (int)$this->faker->boolean($gridFilterProbability);
```
Comment (aRn0D) [inline Generator/AttributeGenerator.php]: missing space

File: Generator/AttributeGroupGenerator.php [hunk 11]
```diff
@@ -20,32 +21,37 @@ class AttributeGroupGenerator implements GeneratorInterface
...
     const ATTR_GROUP_CODE_PREFIX = 'attr_gr_';
 
-    const ATTRIBUTE_GROUP_FILENAME = 'attribute_groups.yml';
+    const ATTRIBUTE_GROUP_FILENAME = 'attribute_groups.csv';
+
+    /** @var WriterInterface */
+    protected $writer;
 
     /** @var array */
     protected $attributeGroups;
```
Comment (arnolanglade) [inline Generator/AttributeGroupGenerator.php]: protected $attributeGroups = []; or you should initialize in the constructor

File: Generator/AttributeGroupGenerator.php [hunk 12]
```diff
@@ -20,32 +21,37 @@ class AttributeGroupGenerator implements GeneratorInterface
...
+        $this->writer = $writer;
+    }
+
+    /**
      * {@inheritdoc}
      */
     public function generate(array $globalConfig, array $config, ProgressHelper $progress, array $options = [])
     {
         $this->locales = $options['locales'];
+        $count         = (int)$config['count'];
```
Comment (arnolanglade) [inline Generator/AttributeGroupGenerator.php:52]: missing space

File: Generator/AttributeOptionGenerator.php [hunk 13]
```diff
@@ -19,11 +21,8 @@ class AttributeOptionGenerator implements GeneratorInterface
...
-    /** @var string */
-    protected $outputFile;
-
-    /** @var string */
-    protected $delimiter;
+    /** @var WriterInterface */
+    protected $writer;
 
     /** @var array */
     protected $locales;
```
Comment (arnolanglade) [inline Generator/AttributeOptionGenerator.php]: protected $locales = [] or in constructor;

File: Generator/AttributeOptionGenerator.php [hunk 14]
```diff
@@ -41,6 +40,14 @@ class AttributeOptionGenerator implements GeneratorInterface
     protected $faker;
```
Comment (arnolanglade) [inline Generator/AttributeOptionGenerator.php:43]: You should remove the FQCN

File: Generator/AttributeOptionGenerator.php [hunk 15]
```diff
@@ -73,9 +77,10 @@ public function generate(array $globalConfig, array $config, ProgressHelper $pro
             };
         }
 
-        $headers = $this->getAllKeys($this->attributeOptions);
-
-        $this->writeCsvFile($this->attributeOptions, $headers);
+        $this->writer
+            ->setFilename($globalConfig['output_dir'].'/'.static::ATTRIBUTE_OPTIONS_FILENAME)
```
Comment (arnolanglade) [inline Generator/AttributeOptionGenerator.php]: sprintf ?

File: Generator/CategoryGenerator.php [hunk 16]
```diff
@@ -16,48 +17,57 @@
...
     const CATEGORIES_CODE_PREFIX = 'cat_';
+
     const LABEL_LENGTH = 2;
 
-    /** @var Locale[] */
+    /** @var WriterInterface */
+    protected $writer;
+
+    /** @var LocaleInterface[] */
     protected $locales;
```
Comment (arnolanglade) [inline Generator/CategoryGenerator.php]: idem it is an array

File: Generator/CategoryGenerator.php [hunk 17]
```diff
@@ -16,48 +17,57 @@
...
-
         $rootCategory = $this->generateCategory('master', 'Master Catalog');
         $categories = $this->generateCategories($rootCategory, 1, $countByLevel, $levelMax);
 
         $normalizedCategories = $this->normalizeCategories($categories);
 
-        $outputFile = $globalConfig['output_dir'].'/'.self::CATEGORIES_FILENAME;
-        $this->writeCsvFile($headers, $normalizedCategories, $outputFile, $delimiter);
+        $this->writer
+            ->setFilename($globalConfig['output_dir'].'/'.self::CATEGORIES_FILENAME)
```
Comment (arnolanglade) [inline Generator/CategoryGenerator.php]: sprintf?

File: Generator/ChannelGenerator.php [hunk 18]
```diff
@@ -17,28 +21,39 @@
...
+    protected $writer;
+
     /** @var string */
     protected $channelsFilePath;
 
     /** @var string */
     protected $currenciesFilePath;
 
-    /** @var Channels[] */
+    /** @var ChannelInterface[] */
```
Comment (arnolanglade) [inline Generator/ChannelGenerator.php:39]: idem

File: Generator/ChannelGenerator.php [hunk 19]
```diff
@@ -203,20 +229,18 @@ protected function writeChannelsFile()
...
-                [
-                    'code'       => $channel->getCode(),
-                    'label'      => $channel->getLabel(),
-                    'tree'       => static::DEFAULT_TREE,
-                    'locales'    => $localeCodes,
-                    'currencies' => $currencyCodes,
-                    'color'      => $channel->getColor()
-                ];
+            $data[] = [
+                'code' => $channel->getCode(),
```
Comment (arnolanglade) [inline Generator/ChannelGenerator.php]: alignement

File: Generator/ChannelGenerator.php [hunk 20]
```diff
@@ -203,20 +229,18 @@ protected function writeChannelsFile()
...
+                'label'      => $channel->getLabel(),
+                'tree'       => static::DEFAULT_TREE,
```
Comment (arnolanglade) [inline Generator/ChannelGenerator.php]: self?

File: Generator/FamilyGenerator.php [hunk 21]
```diff
@@ -42,7 +39,7 @@ class FamilyGenerator implements GeneratorInterface
     /** @var array */
     protected $channels;
```
Comment (arnolanglade) [inline Generator/FamilyGenerator.php]: initialization !

File: Generator/FamilyGenerator.php [hunk 22]
```diff
@@ -111,9 +109,10 @@ public function generate(array $globalConfig, array $config, ProgressHelper $pro
 
         $this->families = $families;
 
-        $headers = $this->getAllKeys($this->families);
-
-        $this->writeCsvFile($this->families, $headers);
+        $this->writer
+            ->setFilename($globalConfig['output_dir'].'/'.self::FAMILIES_FILENAME)
```
Comment (arnolanglade) [inline Generator/FamilyGenerator.php]: sprintf ?

File: Generator/JobGenerator.php [hunk 23]
```diff
@@ -127,7 +127,7 @@ protected function getInternalJobs()
         $internalJobsPath = __DIR__.'/../'.static::INTERNAL_JOBS_FILE;
```
Comment (arnolanglade) [inline Generator/JobGenerator.php]: sprintf ?

File: Generator/JobGenerator.php [hunk 24]
```diff
@@ -127,7 +127,7 @@ protected function getInternalJobs()
...
         $yamlParser = new Yaml\Parser();
```
Comment (arnolanglade) [inline Generator/JobGenerator.php]: you should use this class

File: Generator/JobProfilesAccessGenerator.php [hunk 25]
```diff
@@ -16,57 +18,52 @@
...
-     * @param array  $data
-     * @param string $outputDir
-     */
-    protected function writeYamlFile(array $data, $outputDir)
-    {
-        $dumper = new Yaml\Dumper();
-        $yamlData = $dumper->dump($data, 3, 0, true, true);
-
-        file_put_contents($outputDir.'/'.self::JOB_PROFILE_ACCESSES_FILENAME, $yamlData);
+        $this->writer
```
Comment (arnolanglade) [inline Generator/JobProfilesAccessGenerator.php:62]: sprintf ?

File: Generator/LocaleGenerator.php [hunk 26]
```diff
@@ -0,0 +1,31 @@
+<?php
+
+namespace Pim\Bundle\DataGeneratorBundle\Generator;
+use Symfony\Component\Console\Helper\ProgressHelper;
```
Comment (arnolanglade) [inline Generator/LocaleGenerator.php:5]: extra line

File: Generator/LocaleGenerator.php [hunk 27]
```diff
@@ -0,0 +1,31 @@
...
+
+    const INTERNAL_LOCALES_FILE = 'Resources/config/locales.csv';
+
+    /**
+     * {@inheritdoc}
+     */
+    public function generate(array $globalConfig, array $generatorConfig, ProgressHelper $progress, array $options = [])
+    {
+        copy(
+            dirname(__FILE__) . '/../' . self::INTERNAL_LOCALES_FILE,
```
Comment (arnolanglade) [inline Generator/LocaleGenerator.php]: sprintf?

File: Generator/UserGenerator.php [hunk 28]
```diff
@@ -19,9 +21,12 @@
...
+class UserGenerator implements GeneratorInterface
 {
-    const USERS_FILENAME = 'users.yml';
+    const USERS_FILENAME = 'users.csv';
+
+    /** @var WriterInterface */
+    protected $writer;
 
     /** @var Channel[] */
     protected $channels;
```
Comment (arnolanglade) [inline Generator/UserGenerator.php]: array initialization ?

File: Generator/UserGenerator.php [hunk 29]
```diff
@@ -64,10 +77,10 @@ public function generate(array $globalConfig, array $config, ProgressHelper $pro
 
         $normalizedUsers = $this->normalizeUsers($users);
 
-        $this->writeYamlFile(
-            $normalizedUsers,
-            $globalConfig['output_dir']. "/" . static::USERS_FILENAME
-        );
+        $this->writer
+            ->setFilename($globalConfig['output_dir']. "/" . self::USERS_FILENAME)
```
Comment (arnolanglade) [inline Generator/UserGenerator.php]: sprintf?

File: Generator/UserGroupGenerator.php [hunk 30]
```diff
@@ -24,12 +38,12 @@ public function generate(array $globalConfig, array $config, ProgressHelper $pro
...
 
-        $normalizedGroups = $this->normalizeGroups($groups);
+        $normalizedGroups = $this->normalizeGroups(array_values($groups));
 
-        $this->writeYamlFile(
-            $normalizedGroups,
-            $globalConfig['output_dir'] . "/" . static::GROUPS_FILENAME
-        );
+        $this->writer
+            ->setFilename($globalConfig['output_dir'] . "/" . static::GROUPS_FILENAME)
```
Comment (arnolanglade) [inline Generator/UserGroupGenerator.php]: sprintf?

File: Generator/JobGenerator.php [hunk 31]
```diff
@@ -124,10 +125,10 @@ protected function normalizeJob(JobInstance $job)
      */
     protected function getInternalJobs()
     {
-        $internalJobsPath = __DIR__.'/../'.static::INTERNAL_JOBS_FILE;
-        $yamlParser = new Yaml\Parser();
+        $internalJobsPath = __DIR__.'/../'.self::INTERNAL_JOBS_FILE;
```
Comment (arnolanglade) [inline Generator/JobGenerator.php]: sprintf ?

**Comment by arnolanglade** [comment]
GTM once fixed (missing sprintf ;))

File: Generator/UserRoleGenerator.php [hunk 32]
```diff
@@ -57,7 +68,7 @@ protected function generateRoles(array $rolesConfig)
      * Generate a role object from the data provided
      *
      * @param string $key
-     * @param array $config
+     * @param array  $roleConfig
```
Comment (pierallard) [inline Generator/UserRoleGenerator.php]: Space

File: Generator/JobProfilesAccessGenerator.php (hunk 25 above)
Comment (pierallard) [inline Generator/JobProfilesAccessGenerator.php:62]: done

File: Generator/AttributeOptionGenerator.php (hunk 14 above)
Comment (pierallard) [inline Generator/AttributeOptionGenerator.php:43]: done

File: Generator/AttributeGenerator.php (hunk 9 above)
Comment (pierallard) [inline Generator/AttributeGenerator.php:146]: I like my var name, should discuss IRL twoomoowroh

File: Generator/JobGenerator.php (hunk 31 above)
Comment (nidup) [inline Generator/JobGenerator.php]: use DIRECTORY_SEPARATOR and not '/'

File: Generator/AssetCategoryAccessGenerator.php [hunk 33]
```diff
@@ -15,55 +16,44 @@
...
-            $data[$assetCategoryCode] = [];
-            foreach (['viewItems', 'editItems'] as $access) {
-                $data[$assetCategoryCode][$access] = [];
-                foreach ($this->groups as $group) {
-                    if ('all' !== $group->getName()) {
-                        $data[$assetCategoryCode][$access][] = $group->getName();
-                    }
-                }
+        $groupNames = [];
+        /** @var Group $group */
```
Comment (solivier) [inline Generator/AssetCategoryAccessGenerator.php]: nope remove this line useless

File: Resources/examples/fixtures.yml [hunk 34]
```diff
@@ -50,17 +50,21 @@ data_generator:
             count: 50
             levels: 3
         jobs:
-            csv_product_import:
-                connector: Akeneo CSV Connector
-                alias: csv_product_import
-                label: Demo product import
-                type: import
+            product_import:
```
Comment (nidup) [inline Resources/examples/fixtures.yml]: why do we change the job code?

**Comment by nidup** [comment]
baby comment then GTM

File: Generator/AttributeGenerator.php [hunk 35]
```diff
@@ -280,9 +277,11 @@ protected function getAttributeGroupCodes()
     protected function getLocalizedRandomLabels($type)
     {
         $labels = [];
+        $smallType = str_replace('pim_catalog_', '', $type);
 
+        /** @var LocaleInterface $locale */
```
Comment (solivier) [inline Generator/AttributeGenerator.php]: you can remove this line, it's useless you added it only for your ide

File: Generator/AttributeGroupsAccessGenerator.php [hunk 36]
```diff
@@ -17,56 +19,50 @@
...
-            $data[$attributeGroupCode] = [];
-            foreach (['viewAttributes', 'editAttributes'] as $access) {
-                $data[$attributeGroupCode][$access] = [];
-                foreach ($this->groups as $group) {
-                    if ('all' !== $group->getName()) {
-                        $data[$attributeGroupCode][$access][] = $group->getName();
-                    }
-                }
+        $groupNames = [];
+        /** @var Group $group */
```
Comment (solivier) [inline Generator/AttributeGroupsAccessGenerator.php]: remove this line

File: Generator/AttributeGroupsAccessGenerator.php [hunk 37]
```diff
@@ -17,56 +19,50 @@
...
+        foreach ($groups as $group) {
+            if (User::GROUP_DEFAULT !== $group->getName()) {
+                $groupNames[] = $group->getName();
             }
         }
 
-        $attributeGroupsAccesses = [self::ATTRIBUTE_GROUPS_ACCESSES => $data];
-
+        /** @var AttributeGroupInterface $attributeGroup */
```
Comment (solivier) [inline Generator/AttributeGroupsAccessGenerator.php]: remove the line

File: DependencyInjection/PimDataGeneratorExtension.php [hunk 38]
```diff
@@ -22,5 +22,6 @@ public function load(array $configs, ContainerBuilder $container)
         $loader = new YamlFileLoader($container, new FileLocator(__DIR__.'/../Resources/config'));
```
Comment (nidup) [inline DependencyInjection/PimDataGeneratorExtension.php]: use DIRECTORY_SEPARATOR constant please

**Pull request merged**